pkg/virtctl/scp/scp.go

```diff
@@ -0,0 +1,181 @@
+		},
+	}
+	cmd.Flags().BoolVarP(&c.recursive, "recursive", "r", c.recursive, "Recursively copy entire directories")
+	cmd.Flags().BoolVar(&c.preserve, "preserve", c.preserve, "Preserves modification times, access times, and modes from the original file.")
+	ssh.AddCommandlineArgs(cmd.Flags(), &c.options)
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+	return cmd
+}
+
+type SCP struct {
+	clientConfig clientcmd.ClientConfig
+	options      ssh.SSHOptions
+	recursive    bool
+	preserve     bool
+}
+
+func (o *SCP) Run(cmd *cobra.Command, args []string) error {
+	local, remote, toRemote, err := PrepareCommand(cmd, o.clientConfig, &o.options, args)
+	if err != nil {
+		return err
+	}
+
+	sshClient := ssh.NativeSSHConnection{
+		ClientConfig: o.clientConfig,
+		Options:      o.options,
+	}
+	client, err := sshClient.PrepareSSHClient(remote.Kind, remote.Namespace, remote.Name)
+	if err != nil {
+		return err
+	}
+	scpClient, err := scp.NewClientFromExistingSSH(client, &scp.ClientOption{})
+	if err != nil {
+		return err
+	}
+	if toRemote {
+		if o.recursive {
+			stat, err := os.Stat(local.Path)
+			if err != nil {
+				return fmt.Errorf("failed reading path %q: %v", local.Path, err)
+			}
+			if !stat.IsDir() {
+				return fmt.Errorf("local path %q is not a directory", local.Path)
+			}
+			err = scpClient.CopyDirToRemote(local.Path, remote.Path, &scp.DirTransferOption{PreserveProp: o.preserve})
+			if err != nil {
+				return err
+			}
+		} else {
+			err = scpClient.CopyFileToRemote(local.Path, remote.Path, &scp.FileTransferOption{PreserveProp: o.preserve})
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		if o.recursive {
+			stat, err := os.Stat(local.Path)
+			if !os.IsNotExist(err) {
+				if err != nil {
+					// stat failed, let's abort on weird errors
+					return fmt.Errorf("failed reading path %q: %v", local.Path, err)
+				} else if !stat.IsDir() {
+					// if the local path exists already, it must be a directory
+					return fmt.Errorf("local path %q is not a directory", local.Path)
+				}
+				// the local directory already exists, let's check if we want to create a sub-directory
+				// with the remote base name.
+				remoteBase := filepath.Base(remote.Path)
+				switch remoteBase {
+				case "..", ".", "/", "./", "":
+					// no identifiable base name, let's go with the supplied local path
+				default:
+					// we identified a base location, let's append it to the local path
+					local.Path = filepath.Join(local.Path, remoteBase)
+				}
+			}
+
+			if err := os.MkdirAll(local.Path, os.ModePerm); err != nil {
+				return fmt.Errorf("failed ensuring the existence of the local target directory %q: %v", local.Path, err)
+			}
+
+			err = scpClient.CopyDirFromRemote(remote.Path, local.Path, &scp.DirTransferOption{PreserveProp: o.preserve})
+			if err != nil {
+				return err
+			}
+		} else {
+			err = scpClient.CopyFileFromRemote(remote.Path, local.Path, &scp.FileTransferOption{PreserveProp: o.preserve})
```

Comment:
Same as above. What happens if `remote.Path` is a dir?

Reply:
It wil fail with `no such file or directory.

Reply:
I would find this error confusing since it is actually a directory.

Reply:
sorry I was not clear enough. If it does not exist it will tell you that there is no file or directory. Otherwise it will tell you that it is 'not a regular file'.